man/_ash_log.1
.\"
.\"Copyright 2016 Carl Anderson
.\"
.\"Licensed under the Apache License, Version 2.0 (the "License");
.\"you may not use this file except in compliance with the License.
.\"You may obtain a copy of the License at
.\"
.\"    http://www.apache.org/licenses/LICENSE-2.0
.\"
.\"Unless required by applicable law or agreed to in writing, software
.\"distributed under the License is distributed on an "AS IS" BASIS,
.\"WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.\"See the License for the specific language governing permissions and
.\"limitations under the License.
.\"

.TH _ash_log 1 \
  "Updated: __DATE__" \
  "__VERSION__" \
  "Advanced Shell History"


.SH NAME
_ash_log - The advanced shell history command logger.


.SH SYNOPSIS
Usage: _ash_log [options]
  -h  --help
  -a  --alert VALUE
  -c  --command VALUE
  -e  --command_exit VALUE
  -p  --command_pipe_status VALUE
  -s  --command_start VALUE
  -f  --command_finish VALUE
  -n  --command_number VALUE
  -x  --exit VALUE
  -V  --version
  -S  --get_session_id
  -E  --end_session


.SH DESCRIPTION
.B _ash_log and _ash_log.py are not intended to be executed manually.
They are invoked automatically by the advanced shell history system which is 
only activated by sourcing the appropriate file (depending on whether you are
using bash or zsh).

Advanced Shell History includes:
.RS
Session details:
.RS
.IP "* login and logout times"
.IP "* username and hostname"
.IP "* incoming ssh connection details (ip and ports)"
.IP "* shell name"
.IP "* tty, shell pid and ppid"
.RE

Command details:
.RS
.IP "* start and stop times"
.IP "* exit code and pipestatus codes"
.IP "* current working directory"
.IP "* session id"
.RE

Many more data points are also collected.
.RE


If you intend to use this system to manage your own personal command history,
it is recommended that you add the appropriate source command to your shell rc
file so it is automatically started with each new session.

.B For bash, add this to your ~/.bashrc file:
.RS
source /usr/local/lib/advanced_shell_history/sh/bash
.RE

.B For zsh, add this to your ~/.zshrc file:
.RS
source /usr/local/lib/advanced_shell_history/sh/zsh
.RE


.SH OPTIONS
.IP "  -h  --help"

Display command help and exit 0.

.IP "  -a  --alert VALUE"

Display an alert message (VALUE).

.IP "  -c  --command VALUE"

Store the command (VALUE) entered by the user.

.IP "  -e  --command_exit VALUE"

Store the command exit code (VALUE) of the command.

.IP "  -p  --command_pipe_status VALUE"

Store the command pipe states (exit code for each piped command) separated by
underscores.

.IP "  -s  --command_start VALUE"

The unix epoch timestamp (VALUE) when the command started (the user pressed
Enter).

.IP "  -f  --command_finish VALUE"

The unix epoch timestamp (VALUE) when the command completed and the next prompt
was displayed.

.IP "  -n  --command_number VALUE"

The shell builtin history number (VALUE) of the entered command.

.IP "  -x  --exit VALUE"

The exit code to use when exiting this program.
This flag is needed to re-create the same exit status as the logged command.
This is important because this command runs automatically before the prompt is
redrawn, so if _ash_log does not restore the previous exit code, the user would
see the exit code for _ash_log (and NOT their command) when executing:
.RS
  echo ${?}
.RE

.IP "  -V  --version"

Display the version number and exit.

.IP "  -S  --get_session_id"

Display the advanced shell history session number and exit.  If no session ID
is found, one is created and displayed.

.IP "  -E  --end_session"

Ends the current session, as defined by the shell environment variable
ASH_SESSION_ID.  It is an error to use this flag without having the
ASH_SESSION_ID variable set.


.SH FILES
.I /etc/ash/ash.conf
.RS
Contains environment variables to be sourced into your shell.
.RE

.I /etc/ash/queries
.RS
Contains saved queries which can be invoked using the
.BR ash_query(1)
command.  Also see
.I ~/.ash/queries
for user-written queries.
.RE

.I ~/.ash/history.db
.RS
The default location for the sqlite3 database holding command history.  See
https://github.com/barabo/advanced-shell-history/w for more details on how
the data is stored internally and other tips for querying the data.
.RE

.I /usr/local/lib/advanced_shell_history/sh/bash
.RS
Sourced for bash sessions.
.RE

.I /usr/local/lib/advanced_shell_history/sh/zsh
.RS
Sourced for zsh sessions.
.RE


.SH ENVIRONMENT
.IP ASH_CFG_DB_FAIL_RANDOM_TIMEOUT
After a failed insert, sleep a random number of milliseconds before retrying.
This is intended to add some noise to the retry mechanism.

.IP ASH_CFG_DB_FAIL_TIMEOUT
After a failed insert, sleep this many milliseconds before retrying.

.IP ASH_CFG_DB_MAX_RETRIES
Quit db retries after this many failed attempts.

.IP ASH_CFG_HIDE_USAGE_FOR_NO_ARGS
Normally, if you invoke ash_query with no arguments, the --help output is
displayed.  With this set to a non-empty value, the --help output is
suppressed in this case.

.IP ASH_CFG_HISTORY_DB
The default database to query.  This is set by sourcing one of the shell
scripts in /usr/local/lib/advanced_shell_history/sh and signifies the location
of the database where commands are logged.  If this variable exists, the
--database flag does not need to be used.

.IP ASH_CFG_IGNORE_UNKNOWN_FLAGS
Normally ash_query complains when it sees unknown flags.  With this variable
set to a non-empty value, unknown flags are ignored.

.IP ASH_CFG_LOG_DATE_FMT
If logging is in use, this format string can be set to customize the date
string.

.IP ASH_CFG_LOG_FILE
The file destination of logged messages, if logging is in use.

.IP ASH_CFG_LOG_IPV4
Can be used to skip logging ipv4 host IP addresses.

.IP ASH_CFG_LOG_IPV6
Can be used to skip logging ipv6 host IP addresses.

.IP ASH_CFG_LOG_LEVEL
The lowest level of logging to make visible.  Levels (in increasing order)
are DEBUG, INFO, WARN, ERROR and FATAL.

.IP ASH_CFG_SKIP_LOOPBACK
Skip logging IP addresses for loopback devices (both ipv4 and ipv6).

.IP ASH_DISABLED
If set, _ash_log is disabled.

.IP ASH_LOG_BIN
The binary used to log command history to the database.  This should be either
_ash_log or _ash_log.py, depending on your system setup.

.IP ASH_SESSION_ID
The session id number created for the current session.  If unset, a new
session ID will be created and displayed.  The caller is expected to export
this variable using the generated ID number.


.SH "SEE ALSO"
.BR ash_query(1)
to query history


.SH AUTHOR
Carl Anderson, Mayo Clinic, Inc.


.SH BUGS
Report bugs at https://github.com/barabo/advanced-shell-history/issues


python/_ash_log.py
#!/usr/bin/python
#
# Copyright 2018 Carl Anderson
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""A module to log commands and shell session details into a sqlite3 database.

This module logs a command into a command history database, while also gathering
system-specific metadata.
"""
from __future__ import print_function

__author__ = 'Carl Anderson (carl.anderson@gmail.com)'
__version__ = '0.9r0'


import logging
import os
import sys

# Allow the local advanced_shell_history library to be imported.
_LIB = '/usr/local/lib'
if _LIB not in sys.path:
  sys.path.append(_LIB)

from advanced_shell_history import unix
from advanced_shell_history import util


class Flags(util.Flags):
  """The flags needed for the _ash_log.py script to work."""

  arguments = (
    ('a', 'alert', 'MSG', str, 'a message to display to the user'),
    ('c', 'command', 'CMD', str, 'a command to log'),
    ('e', 'command_exit', 'CODE', int, 'the exit code of the command to log'),
    ('p', 'command_pipe_status', 'CSV', str, 'the pipe states of the command to log'),
    ('s', 'command_start', 'TS', int, 'the timestamp when the command started'),
    ('f', 'command_finish', 'TS', int, 'the timestamp when the command stopped'),
    ('n', 'command_number', 'NUM', int, 'the builtin shell history command number'),
    ('x', 'exit', 'CODE', int, 'the exit code to use when exiting'),
  )

  flags = (
    ('S', 'get_session_id', 'emits the session ID (or creates one)'),
    ('E', 'end_session', 'ends the current session'),
  )

  def __init__(self):
    util.Flags.__init__(self, Flags.arguments, Flags.flags)


class Session(util.Database.Object):
  """An abstraction of a shell session to store to the history database."""

  def __init__(self):
    """Initialize a Session, populating session values."""
    util.Database.Object.__init__(self, 'sessions')
    self.values = {
      'time_zone': unix.GetTimeZone(),
      'start_time': unix.GetTime(),
      'ppid': unix.GetPPID(),
      'pid': unix.GetPID(),
      'tty': unix.GetTTY(),
      'uid': unix.GetUID(),
      'euid': unix.GetEUID(),
      'logname': unix.GetLoginName(),
      'hostname': unix.GetHostName(),
      'host_ip': unix.GetHostIp(),
      'shell': unix.GetShell(),
      'sudo_user': unix.GetEnv('SUDO_USER'),
      'sudo_uid': unix.GetEnv('SUDO_UID'),
      'ssh_client': unix.GetEnv('SSH_CLIENT'),
      'ssh_connection': unix.GetEnv('SSH_CONNECTION')
    }

  def GetCreateTableSql(self):
    return '''
CREATE TABLE sessions ( 
  id integer primary key autoincrement, 
  hostname varchar(128), 
  host_ip varchar(40), 
  ppid int(5) not null, 
  pid int(5) not null, 
  time_zone str(3) not null, 
  start_time integer not null, 
  end_time integer, 
  duration integer, 
  tty varchar(20) not null, 
  uid int(16) not null, 
  euid int(16) not null, 
  logname varchar(48), 
  shell varchar(50) not null, 
  sudo_user varchar(48), 
  sudo_uid int(16), 
  ssh_client varchar(60), 
  ssh_connection varchar(100) 
)'''

  def Close(self):
    """Closes this session in the database."""
    sql = '''
      UPDATE sessions
      SET
        end_time = ?,
        duration = ? - start_time
      WHERE id == ?;
    '''
    ts = unix.GetTime()
    util.Database().Execute(sql, (ts, ts, unix.GetEnvInt('ASH_SESSION_ID'),))


class Command(util.Database.Object):
  """An abstraction of a command to store to the history database."""
  def __init__(self, command, rval, start, finish, number, pipes):
    util.Database.Object.__init__(self, 'commands')
    self.values = {
      'session_id': unix.GetEnvInt('ASH_SESSION_ID'),
      'shell_level': unix.GetEnvInt('SHLVL'),
      'command_no': number,
      'tty': unix.GetTTY(),
      'euid': unix.GetEUID(),
      'cwd': unix.GetCWD(),
      'rval': rval,
      'start_time': start,
      'end_time': finish,
      'duration': finish - start,
      'pipe_cnt': len(pipes.split('_')),
      'pipe_vals': pipes,
      'command': command
    }
    # If the user changed directories, CWD will be the new directory, not the
    # one where the command was actually entered.
    if rval == 0 and (command == 'cd' or command.startswith('cd ')):
      self.values['cwd'] = unix.GetEnv('OLDPWD')

  def GetCreateTableSql(self):
    return '''
CREATE TABLE commands (
  id integer primary key autoincrement,
  session_id integer not null,
  shell_level integer not null,
  command_no integer,
  tty varchar(20) not null,
  euid int(16) not null,
  cwd varchar(256) not null,
  rval int(5) not null,
  start_time integer not null,
  end_time integer not null,
  duration integer not null,
  pipe_cnt int(3),
  pipe_vals varchar(80),
  command varchar(1000) not null,
UNIQUE(session_id, command_no)
)'''


def main(argv):
  # If ASH_DISABLED is set, we skip everything and exit without error.
  if os.getenv('ASH_DISABLED'): return 0

  # Setup.
  util.InitLogging()

  # Log the command, if debug logging is enabled.
  if logging.getLogger().isEnabledFor(logging.DEBUG):
    command = []
    for arg in argv:
      command.append('[%d]=\'%s\'' % (len(command), arg))
    logging.debug('argv = "' + ','.join(command) + '"')

  # Print an alert if one was specified.
  flags = Flags()
  if flags.alert:
    print(flags.alert, file=sys.stderr)

  # If no arguments were given, it may be best to show --help.
  if len(argv) == 1 and not util.Config().GetBool('HIDE_USAGE_FOR_NO_ARGS'):
    flags.PrintHelp()

  # Create the session id, if not already set in the environment.
  session_id = os.getenv('ASH_SESSION_ID')
  if flags.get_session_id:
    if session_id is None:
      session_id = Session().Insert()
    print(session_id)

  # Insert a new command into the database, if one was supplied.
  command_flag_used = bool(flags.command
    or flags.command_exit
    or flags.command_pipe_status
    or flags.command_start
    or flags.command_finish
    or flags.command_number)
  if command_flag_used:
    Command(
      flags.command, flags.command_exit, flags.command_start,
      flags.command_finish, flags.command_number, flags.command_pipe_status
    ).Insert()

  # End the current session.
  if flags.end_session:
    Session().Close()

  # Return the desired exit code.
  return flags.exit


if __name__ == '__main__':
  sys.exit(main(sys.argv))
